go-controller/pkg/ovn/namespace.go

```diff
@@ -218,7 +236,66 @@ func (oc *Controller) updateNamespace(old, newer *kapi.Namespace) {
 	}
 	defer nsInfo.Unlock()
 
-	annotation := newer.Annotations[hotypes.HybridOverlayExternalGw]
+	var annotation, oldAnnotation string
+	annotation = newer.Annotations[routingExternalGWsAnnotation]
+	oldAnnotation = old.Annotations[routingExternalGWsAnnotation]
+	if annotation != oldAnnotation {
+		var stderr string
+		var err error
+		for podIP, gwToGr := range nsInfo.podExternalRoutes {
+			for gw, gr := range gwToGr {
+				// TODO figure out how to delete route with gw
+				_, stderr, err = util.RunOVNNbctl("--if-exists", "--policy=src-ip",
+					"lr-route-del", gr, podIP+"/32", gw)
+			}
+			if err != nil {
+				klog.Errorf("Unable to delete src-ip route to GR router, stderr:%q, err:%v", stderr, err)
+			} else {
+				delete(nsInfo.podExternalRoutes, podIP)
+			}
+		}
+		if annotation != "" {
+			annotationOk := true
+			nsInfo.routingExternalGWs = nil
+			for _, v := range strings.Split(annotation, ",") {
+				parsedAnnotation := net.ParseIP(v)
+				if parsedAnnotation == nil {
+					klog.Errorf("Could not parse routing external gw annotation")
+					annotationOk = false
+					nsInfo.routingExternalGWs = nil
+					break
+				} else {
+					nsInfo.routingExternalGWs = append(nsInfo.routingExternalGWs, parsedAnnotation)
+				}
+			}
+			if annotationOk {
+				existingPods, err := oc.watchFactory.GetPods(old.Name)
+				if err != nil {
+					klog.Errorf("Failed to get all the pods (%v)", err)
+				} else {
+					for _, pod := range existingPods {
+						podIP := pod.Status.PodIP
+						gr := "GR_" + pod.Spec.NodeName
+						for _, gw := range nsInfo.routingExternalGWs {
+							_, stderr, err = util.RunOVNNbctl("--may-exist", "--policy=src-ip", "--ecmp",
+								"lr-route-add", gr, podIP+"/32", gw.String())
```

Comment:
We should not be adding new IPv4-only code (`+"/32"`) to the tree. We _must_ not add new code that _blindly_ assumes single-stack IPv4. You need to do one of (in order of preference):

1. add routing rules for each element of `pod.Status.PodIPs`, not just for `pod.Status.PodIP`, OR
2. examine each element of `pod.Status.PodIPs` to find which one is the IPv4 IP and add a rule only for that IP, rather than just assuming that `pod.Status.PodIP` is IPv4, OR
3. ignore the `routing-external-gws` annotation if the cluster is not single-stack IPv4

(likewise elsewhere)

Reply:
ack, will look at option 1

Reply:
I would go with option 1 as well, lets just add support now

Reply:
done